### PR TITLE
saving output from different ranks

### PR DESCRIPTION
**Pull request by hanaol**
Handles logging/saving the performance metric across multiple ranks.

### CODE REVIEW

File: src/electrai/lightning.py [hunk 1]
```diff
@@ -87,6 +87,14 @@ def configure_optimizers(self):
     def on_test_start(self):
         self.log_dir = self.test_cfg.log_dir
         self.out_dir = self.test_cfg.out_dir
+        if self.out_dir is not None:
+            self.out_dir = Path(self.out_dir)
+            self.out_dir.mkdir(exist_ok=True, parents=True)
+        if self.log_dir is not None:
+            self.log_dir = Path(self.log_dir)
+            self.log_dir.mkdir(exist_ok=True, parents=True)
+            self.tmp_dir = Path(self.out_dir) / "tmp"
```
Comment (forklady42) [inline src/electrai/lightning.py]: `self.out_dir` could be `None` here. Should this be in the previous block?

File: src/electrai/lightning.py [hunk 2]
```diff
@@ -110,40 +118,38 @@ def test_step(self, batch, batch_idx):
...
+                np.save(self.out_dir / f"{idx}.npy", preds[i].squeeze(0).cpu().numpy())
 
-        self.test_outputs.append(outputs)
+        if self.log_dir is not None:
+            # Save batch-level CSV
+            if isinstance(nmae, torch.Tensor) and nmae.ndim == 0:
+                nmae = nmae.unsqueeze(0)
+            tmp_csv = self.tmp_dir / f"metrics_batch_{self.global_rank}_{batch_idx}.csv"
+            with open(tmp_csv, "w") as f:
+                for i, n in zip(indices, nmae, strict=False):
```
Comment (forklady42) [inline src/electrai/lightning.py]: Strongly prefer `strict=True` unless there is a reason we do not expect the dimensions to match.

File: src/electrai/lightning.py [hunk 3]
```diff
@@ -110,40 +118,38 @@ def test_step(self, batch, batch_idx):
...
-                f.write("index,nmae\n")
-                for ind, err in zip(index, nmae.tolist(), strict=False):
-                    f.write(f"{ind},{err}\n")
+        # write final CSV with header
+        with open(final_csv, "w") as f_out:
+            f_out.write("index,nmae\n")
+            for tmp_csv in all_tmp_csvs:
+                with open(tmp_csv) as f_in:
+                    for line in f_in:
+                        f_out.write(line)
```
Comment (forklady42) [inline src/electrai/lightning.py]: Would be nice to clean up the `tmp_dir` files once this is done.

**Review by forklady42 — CHANGES_REQUESTED** [review]

File: src/electrai/lightning.py [hunk 4]
```diff
@@ -105,45 +108,45 @@ def test_step(self, batch, batch_idx):
...
+        preds = outputs["pred"]
+        indices = outputs["index"]
+        nmae = outputs["nmae"]
 
+        if self.save_pred:
             for i in range(len(indices)):
                 idx = indices[i]
-                pred_i = preds[i].numpy()
-                np.save(out_dir / f"{idx}.npy", pred_i)
+                np.save(self.out_dir / f"{idx}.npy", preds[i].squeeze(0).cpu().numpy())
```
[the remark this answers is not in the document]
Comment (hanaol) [reply src/electrai/lightning.py]: @forklady42 do we want this to be saved/logged in the metrics (`csv`) file as well?

File: src/electrai/lightning.py [hunk 5]
```diff
@@ -105,45 +108,45 @@ def test_step(self, batch, batch_idx):
...
-            n = o["nmae"]
-            if n.ndim == 0:
-                nmae_all.append(n.unsqueeze(0))
-            else:
-                nmae_all.append(n)
+        final_csv = self.log_dir / "metrics.csv"
+        all_tmp_csvs = sorted(self.tmp_dir.glob("metrics_batch_*.csv"))
 
-        nmae = torch.cat(nmae_all, dim=0)
+        if self.global_rank == 0:
```
[the remark this answers is not in the document]
Comment (hanaol) [reply src/electrai/lightning.py]: sgtm

**Review by hanaol — COMMENTED** [review]
addressed the comments
